Sources/SwiftProtobuf/ProtobufBinaryTypes.swift

```diff
@@ -789,14 +777,21 @@ public extension ProtobufBinaryMessageBase {
 
 public extension ProtobufMessage {
     static func decodeProtobufMapValue(decoder: inout ProtobufFieldDecoder, value: inout Self?) throws {
-        let handled = try decoder.decodeSingularMessageField(fieldType: Self.self, value: &value)
-        assert(handled)
+        try decoder.decodeSingularMessageField(fieldType: Self.self, value: &value)
+        assert(value != nil)
     }
 
     init(protobuf: Data) throws {
         try self.init(protobuf: protobuf, extensions: nil)
     }
 
+    init(protobufBuffer: UnsafeBufferPointer<UInt8>) throws {
+        try self.init(protobufBuffer: protobufBuffer, extensions: nil)
+    }
+}
+
+// For Proto3 messages, we need to preserve the unknown fields when we're done
```

Comment:
> At least two of the language implementations (Obj-C and Java) don't bother distinguishing between proto2 and proto3 at all in the public message API (they just return null/empty extension sets and unknown stores for proto3, for example),...

That's partly a difference in language style:  Obj-C and Java also both publish hash value and equality tests for all objects, unlike Swift.  Swift tends to prefer publishing this kind of distinction at compile time instead of run time.